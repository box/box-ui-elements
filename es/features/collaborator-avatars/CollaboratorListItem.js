function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import * as React from 'react';
import classnames from 'classnames';
import { FormattedMessage } from 'react-intl';
import { Link } from '../../components/link';
import PlainButton from '../../components/plain-button';
import Tooltip from '../../components/tooltip';
import IconClose from '../../icon/fill/X16';
import { COLLAB_GROUP_TYPE, COLLAB_PENDING_TYPE } from './constants';
import messages from './messages';
import commonMessages from '../../elements/common/messages';
import CollaboratorAvatarItem from './CollaboratorAvatarItem';
import './CollaboratorListItem.scss';
const CollaboratorListItem = props => {
  const {
    index,
    trackingProps,
    canRemoveCollaborators = false,
    onRemoveCollaborator
  } = props;
  const {
    usernameProps,
    emailProps
  } = trackingProps;
  const {
    email,
    expiration,
    hasCustomAvatar,
    name,
    type,
    imageURL,
    isExternalCollab,
    profileURL,
    translatedRole,
    userID,
    isRemovable = false
  } = props.collaborator;
  const userOrGroupNameContent = type !== COLLAB_GROUP_TYPE ? /*#__PURE__*/React.createElement("div", {
    className: classnames('name', type)
  }, /*#__PURE__*/React.createElement(Link, _extends({
    href: profileURL || `/profile/${userID}`,
    rel: "noopener",
    target: "_blank"
  }, usernameProps), name)) : /*#__PURE__*/React.createElement("div", {
    className: classnames('name', type)
  }, name);
  const emailContent = type !== COLLAB_GROUP_TYPE && email ? /*#__PURE__*/React.createElement("div", {
    className: "email"
  }, /*#__PURE__*/React.createElement(Link, _extends({
    href: `mailto:${email}`
  }, emailProps), email)) : null;
  const roleNodeContent = /*#__PURE__*/React.createElement("div", {
    className: "role"
  }, type === COLLAB_PENDING_TYPE ? /*#__PURE__*/React.createElement(FormattedMessage, messages.pendingCollabText) : translatedRole);
  return /*#__PURE__*/React.createElement("li", null, /*#__PURE__*/React.createElement("div", {
    className: "collaborator-list-item"
  }, /*#__PURE__*/React.createElement("div", {
    className: "bdl-CollaboratorListItem-user user"
  }, /*#__PURE__*/React.createElement("div", {
    className: "info"
  }, userOrGroupNameContent, emailContent), /*#__PURE__*/React.createElement(CollaboratorAvatarItem, {
    allowBadging: true,
    avatarUrl: imageURL,
    email: email,
    expiration: expiration,
    hasCustomAvatar: hasCustomAvatar,
    id: index,
    isExternalCollab: isExternalCollab,
    name: name
  })), canRemoveCollaborators ? /*#__PURE__*/React.createElement("div", {
    className: "user-actions"
  }, roleNodeContent, isRemovable && /*#__PURE__*/React.createElement(Tooltip, {
    isTabbable: false,
    text: /*#__PURE__*/React.createElement(FormattedMessage, commonMessages.remove)
  }, /*#__PURE__*/React.createElement(PlainButton, {
    className: "remove-button",
    onClick: () => onRemoveCollaborator?.(props.collaborator),
    type: "button"
  }, /*#__PURE__*/React.createElement(IconClose, {
    color: "##6f6f6f",
    height: 16,
    width: 16
  })))) : roleNodeContent));
};
export default CollaboratorListItem;
//# sourceMappingURL=CollaboratorListItem.js.map