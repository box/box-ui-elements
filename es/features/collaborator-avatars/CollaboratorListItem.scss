@import '../../styles/variables';

.usm-collaborator-list {
    .collaborator-list-item {
        display: flex;
        justify-content: space-between;
        margin: 10px;

        .user {
            display: flex;
            align-items: center;
            white-space: nowrap;
        }

        .avatar {
            width: 32px;
            height: 32px;
            margin-right: 8px;
        }

        .name {
            max-width: 200px;
            overflow: hidden;
            color: $bdl-gray;
            font-weight: bold;
            font-size: 13px;
            line-height: 16px;
            text-overflow: ellipsis;
        }

        .name .link {
            color: $bdl-gray;
        }

        .email .link {
            color: $bdl-gray-65;
        }

        .name .link:hover,
        .email .link:hover {
            text-decoration: underline;
        }

        .name.group {
            line-height: 32px;
        }

        .email {
            max-width: 200px;
            overflow: hidden;
            color: $bdl-gray-65;
            font-size: 12px;
            line-height: 16px;
            text-overflow: ellipsis;
        }

        .user-actions {
            display: flex;
            gap: 8px;
            align-items: center;
            margin-left: auto;
        }

        .role {
            color: $bdl-gray-65;
            line-height: 32px;
        }

        .remove-button {
            display: flex;
            align-items: center;
            justify-content: center;
            width: $bdl-btn-height;
            height: $bdl-btn-height;
            background: none;
            border: 0;
            cursor: pointer;

            &:hover {
                background: $bdl-gray-05;
                border-radius: $bdl-border-radius-size;
            }
        }
    }

    .collaborator-list-item.more {
        justify-content: inherit;

        .link {
            color: $bdl-box-blue;
            line-height: 32px;
            text-decoration: none;
        }

        .collaboration-badge {
            margin-right: 8px;
        }
    }
}

.bdl-CollaboratorListItem-user {
    flex-direction: row-reverse;
}
