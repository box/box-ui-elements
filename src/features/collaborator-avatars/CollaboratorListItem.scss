@import '../../styles/variables';

.usm-collaborator-list {
    .collaborator-list-item {
        display: flex;
        justify-content: space-between;
        margin: 10px;

        .user {
            display: flex;
            align-items: center;
            white-space: nowrap;
        }

        .avatar {
            width: 32px;
            height: 32px;
            margin-right: 8px;
        }

        .name {
            max-width: 200px;
            overflow: hidden;
            color: $bdl-gray;
            font-weight: bold;
            font-size: 13px;
            line-height: 16px;
            text-overflow: ellipsis;
        }

        .name .link {
            color: $bdl-gray;
        }

        .email .link {
            color: $bdl-gray-62;
        }

        .name .link:hover,
        .email .link:hover {
            text-decoration: underline;
        }

        .name.group {
            line-height: 32px;
        }

        .email {
            max-width: 200px;
            overflow: hidden;
            color: $bdl-gray-62;
            font-size: 12px;
            line-height: 16px;
            text-overflow: ellipsis;
        }

        .role {
            color: $bdl-gray-62;
            line-height: 32px;
        }
    }

    .collaborator-list-item.more {
        justify-content: inherit;

        .link {
            color: $bdl-box-blue;
            line-height: 32px;
            text-decoration: none;
        }

        .collaboration-badge {
            margin-right: 8px;
        }
    }
}
