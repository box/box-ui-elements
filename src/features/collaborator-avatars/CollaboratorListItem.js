// @flow
import React from 'react';
import classnames from 'classnames';
import { FormattedMessage } from 'react-intl';

import { Link } from '../../components/link';

import { COLLAB_GROUP_TYPE, COLLAB_PENDING_TYPE } from './constants';
import messages from './messages';
import CollaboratorAvatarItem from './CollaboratorAvatarItem';
import './CollaboratorListItem.scss';

type Props = {
    collaborator: Object,
    index: number,
    trackingProps: { emailProps: ?Object, usernameProps: ?Object },
};

const CollaboratorListItem = (props: Props) => {
    const { index, trackingProps } = props;
    const { usernameProps, emailProps } = trackingProps;
    const {
        email,
        expiration,
        hasCustomAvatar,
        name,
        type,
        imageURL,
        isExternalCollab,
        profileURL,
        translatedRole,
        userID,
    } = props.collaborator;

    const userOrGroupNameContent =
        type !== COLLAB_GROUP_TYPE ? (
            <div className={classnames('name', type)}>
                <Link
                    href={profileURL || `/profile/${userID}`}
                    rel="noopener"
                    target="_blank"
                    title={name}
                    {...usernameProps}
                >
                    {name}
                </Link>
            </div>
        ) : (
            <div className={classnames('name', type)}>{name}</div>
        );

    const emailContent =
        type !== COLLAB_GROUP_TYPE && email ? (
            <div className="email">
                <Link href={`mailto:${email}`} title={email} {...emailProps}>
                    {email}
                </Link>
            </div>
        ) : null;

    return (
        <li>
            <div className="collaborator-list-item">
                <div className="user">
                    <CollaboratorAvatarItem
                        allowBadging
                        avatarUrl={imageURL}
                        email={email}
                        expiration={expiration}
                        hasCustomAvatar={hasCustomAvatar}
                        id={index}
                        isExternalCollab={isExternalCollab}
                        name={name}
                    />
                    <div className="info">
                        {userOrGroupNameContent}
                        {emailContent}
                    </div>
                </div>
                <div className="role">
                    {type === COLLAB_PENDING_TYPE ? (
                        <FormattedMessage {...messages.pendingCollabText} />
                    ) : (
                        translatedRole
                    )}
                </div>
            </div>
        </li>
    );
};

export default CollaboratorListItem;
