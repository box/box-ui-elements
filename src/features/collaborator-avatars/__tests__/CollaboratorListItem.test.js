import * as React from 'react';

import CollaboratorListItem from '../CollaboratorListItem';

describe('features/collaborator-avatars/CollaboratorListItem', () => {
    const collaborator = {
        name: 'test c',
        email: 'testc@example.com',
        profileUrl: 'http://foo.bar.profile',
        hasCustomAvatar: true,
        imageUrl: 'https://foo.bar',
        expiration: {
            expiresAt: 'Jan 1, 1966',
        },
    };

    const getWrapper = (props = {}) =>
        shallow(
            <CollaboratorListItem
                collaborator={collaborator}
                id="111"
                index={1}
                trackingProps={{
                    usernameProps: undefined,
                    emailProps: undefined,
                }}
                {...props}
            />,
        );

    describe('render()', () => {
        test('should render default component', () => {
            const wrapper = getWrapper();
            expect(wrapper).toMatchSnapshot();
        });

        test('should render pending collaborator', () => {
            const wrapper = getWrapper({
                collaborator: {
                    name: 'test c',
                    email: 'testc@example.com',
                    profileUrl: 'http://foo.bar.profile',
                    hasCustomAvatar: true,
                    imageUrl: 'https://foo.bar',
                    expiration: {
                        expiresAt: 'Jan 1, 1966',
                    },
                    type: 'pending',
                },
            });

            expect(wrapper).toMatchSnapshot();
        });

        test('should render group collaborator', () => {
            const wrapper = getWrapper({
                collaborator: {
                    name: 'test c',
                    email: 'testc@example.com',
                    profileUrl: 'http://foo.bar.profile',
                    hasCustomAvatar: true,
                    imageUrl: 'https://foo.bar',
                    expiration: {
                        expiresAt: 'Jan 1, 1966',
                    },
                    type: 'group',
                },
            });

            expect(wrapper).toMatchSnapshot();
        });

        test('should render component when canRemoveCollaborators prop is true and collaborator is not removable', () => {
            const wrapper = getWrapper({ canRemoveCollaborators: true, collaborator });

            expect(wrapper.find('.role').exists()).toBe(true);
            expect(wrapper.find('PlainButton')).toHaveLength(0);
            expect(wrapper).toMatchSnapshot();
        });

        test('should render component when canRemoveCollaborators prop is true and collaborator is removable', () => {
            const wrapper = getWrapper({
                canRemoveCollaborators: true,
                collaborator: { ...collaborator, isRemovable: true },
            });

            expect(wrapper.find('.role').exists()).toBe(true);
            expect(wrapper.find('PlainButton').exists()).toBe(true);
            expect(wrapper.find('Tooltip').exists()).toBe(true);
            expect(wrapper).toMatchSnapshot();
        });

        test('should call onRemoveCollaborator when onRemoveCollaborator prop is passed', () => {
            const onRemoveCollaboratorMock = jest.fn();
            const wrapper = getWrapper({
                canRemoveCollaborators: true,
                onRemoveCollaborator: onRemoveCollaboratorMock,
                collaborator: { ...collaborator, isRemovable: true },
            });

            const removeButton = wrapper.find('PlainButton');
            removeButton.simulate('click');

            expect(onRemoveCollaboratorMock).toHaveBeenCalledWith({ ...collaborator, isRemovable: true });
            expect(wrapper).toMatchSnapshot();
        });

        test('should not call onRemoveCollaborator when onRemoveCollaborator prop is undefined', () => {
            const onRemoveCollaboratorMock = jest.fn();

            const wrapper = getWrapper({
                canRemoveCollaborators: true,
                onRemoveCollaborator: undefined,
                collaborator: { ...collaborator, isRemovable: true },
            });

            const removeButton = wrapper.find('PlainButton');
            removeButton.simulate('click');

            expect(onRemoveCollaboratorMock).not.toHaveBeenCalled();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
