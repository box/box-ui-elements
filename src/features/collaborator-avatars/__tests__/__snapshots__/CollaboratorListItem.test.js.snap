// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/collaborator-avatars/CollaboratorListItem render() should render default component 1`] = `
<li>
  <div
    className="collaborator-list-item"
  >
    <div
      className="user"
    >
      <CollaboratorAvatarItem
        allowBadging={true}
        email="testc@example.com"
        expiration={
          Object {
            "expiresAt": "Jan 1, 1966",
          }
        }
        hasCustomAvatar={true}
        id={1}
        name="test c"
      />
      <div
        className="info"
      >
        <div
          className="name"
        >
          <Link
            className=""
            href="/profile/undefined"
            rel="noopener"
            target="_blank"
            title="test c"
          >
            test c
          </Link>
        </div>
        <div
          className="email"
        >
          <Link
            className=""
            href="mailto:testc@example.com"
            title="testc@example.com"
          >
            testc@example.com
          </Link>
        </div>
      </div>
    </div>
    <div
      className="role"
    />
  </div>
</li>
`;

exports[`features/collaborator-avatars/CollaboratorListItem render() should render group collaborator 1`] = `
<li>
  <div
    className="collaborator-list-item"
  >
    <div
      className="user"
    >
      <CollaboratorAvatarItem
        allowBadging={true}
        email="testc@example.com"
        expiration={
          Object {
            "expiresAt": "Jan 1, 1966",
          }
        }
        hasCustomAvatar={true}
        id={1}
        name="test c"
      />
      <div
        className="info"
      >
        <div
          className="name group"
        >
          test c
        </div>
      </div>
    </div>
    <div
      className="role"
    />
  </div>
</li>
`;

exports[`features/collaborator-avatars/CollaboratorListItem render() should render pending collaborator 1`] = `
<li>
  <div
    className="collaborator-list-item"
  >
    <div
      className="user"
    >
      <CollaboratorAvatarItem
        allowBadging={true}
        email="testc@example.com"
        expiration={
          Object {
            "expiresAt": "Jan 1, 1966",
          }
        }
        hasCustomAvatar={true}
        id={1}
        name="test c"
      />
      <div
        className="info"
      >
        <div
          className="name pending"
        >
          <Link
            className=""
            href="/profile/undefined"
            rel="noopener"
            target="_blank"
            title="test c"
          >
            test c
          </Link>
        </div>
        <div
          className="email"
        >
          <Link
            className=""
            href="mailto:testc@example.com"
            title="testc@example.com"
          >
            testc@example.com
          </Link>
        </div>
      </div>
    </div>
    <div
      className="role"
    >
      <FormattedMessage
        defaultMessage="Pending"
        id="boxui.collaboratorAvatars.collaboration.pendingCollabText"
      />
    </div>
  </div>
</li>
`;
